haicrypt/hc_nettle_aes.c

```diff
@@ -0,0 +1,448 @@
+      ||(ctx->mode == HCRYPT_CTX_MODE_AESCTR)) {        /* CTR mode decrypts using encryption methods */
+    if (hcrypt_aes_set_encrypt_key (key, key_len * 8, aes_key)) {
+      HCRYPT_LOG (LOG_ERR, "%s", "AES_set_encrypt_key(sek) failed\n");
+      return (-1);
+    }
+  } else {                      /* Decrypt key */
+    if (hcrypt_aes_set_decrypt_key (key, key_len * 8, aes_key)) {
+      HCRYPT_LOG (LOG_ERR, "%s", "AES_set_decrypt_key(sek) failed\n");
+      return (-1);
+    }
+  }
+  return (0);
+}
+
+static int
+hcNettle_AES_Encrypt (hcrypt_CipherData * cipher_data,
+    hcrypt_Ctx * ctx,
+    hcrypt_DataDesc * in_data, int nbin,
+    void *out_p[], size_t out_len_p[], int *nbout_p)
+{
+  hcNettle_AES_data *aes_data = (hcNettle_AES_data *) cipher_data;
+  unsigned char *out_msg;
+  int out_len = 0;              //payload size
+  int pfx_len;
+
+  ASSERT (NULL != ctx);
+  ASSERT (NULL != aes_data);
+  ASSERT ((NULL != in_data) || (1 == nbin));    //Only one in_data[] supported
+
+  /* 
+   * Get message prefix length
+   * to reserve room for unencrypted message header in output buffer
+   */
+  pfx_len = ctx->msg_info->pfx_len;
+
+  /* Get buffer room from the internal circular output buffer */
+  out_msg = hcNettle_AES_GetOutbuf (aes_data, pfx_len, in_data[0].len);
+
+  if (NULL != out_msg) {
+    switch (ctx->mode) {
+      case HCRYPT_CTX_MODE_AESCTR:     /* Counter mode */
+      {
+        /* Get current key (odd|even) from context */
+        AES_KEY *aes_key = &aes_data->aes_key[hcryptCtx_GetKeyIndex (ctx)];
+        unsigned char ctr[AES_BLOCK_SIZE];
+        unsigned char iv[AES_BLOCK_SIZE];
+
+        /* Get input packet index (in network order) */
+        hcrypt_Pki pki = hcryptMsg_GetPki (ctx->msg_info, in_data[0].pfx, 1);
+
+        /*
+         * Compute the Initial Vector
+         * IV (128-bit):
+         *    0   1   2   3   4   5  6   7   8   9   10  11  12  13  14  15
+         * +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+         * |                   0s                  |      pki      |  ctr  |
+         * +---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+         *                            XOR                         
+         * +---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+         * |                         nonce                         +
+         * +---+---+---+---+---+---+---+---+---+---+---+---+---+---+
+         *
+         * pki    (32-bit): packet index
+         * ctr    (16-bit): block counter
+         * nonce (112-bit): number used once (salt)
+         */
+        memset (&ctr[0], 0, sizeof (ctr));
+        hcrypt_SetCtrIV ((unsigned char *) &pki, ctx->salt, iv);
+
+        /* Encrypt packet payload in output message buffer */
+        ctr_crypt (aes_key,        /* ctx */
+                   aes128_encrypt, /* nettle_cipher_func */
+                   AES_BLOCK_SIZE, /* cipher blocksize */
```

Comment:
I am sorry that I didn't mention this line. I have changed from `128` to `AES_BLOCK_SIZE` which is defined as `16`.